### PR TITLE
Network representation optimizations for JSON encoding

### LINKED ISSUE #216: `AlchemiscaleClient.create_network` scales poorly with increasing `AlchemicalNetwork` size
Currently, the `AlchemiscaleClient.create_network` method calls [`network.to_dict()`](https://github.com/openforcefield/alchemiscale/blob/8e98af36fb8847f909600f92ba0d95172d4d4124/alchemiscale/interface/client.py#L128) before serializing this to JSON and shipping it to the `AlchemiscaleAPI`. This is an inefficient approach that scales very poorly with increasing `AlchemicalNetwork` size, as any `Component`s used by the network are duplicated many times in the resulting object, including potentially large `ProteinComponent`s.

Instead of doing this, we should make use of `network.to_keyed_dict`, and use a method similar to [`gufe.tokenization.get_all_gufe_objs`](https://github.com/OpenFreeEnergy/gufe/blob/ae96d30952a4e000fcb690443a6dbd79ec31756a/gufe/tokenization.py#L804) for decomposing the `AlchemicalNetwork` into its component keyed-dicts, ordered in such a way that they could all be re-assembled, and serializing+shipping these to the `AlchemiscaleAPI` instead.

Ideally, this same approach could be used for other `AlchemiscaleClient` methods to improve their efficiency, such as `AlchemicalNetwork` retrieval.

### PR DESCRIPTION
**Pull request by ianmkenney**
Fixes #216

This PR implements a graph based optimization for representing gufe objects as an order dependent chain of keyed dicts, dramatically reducing memory usage (~ -97% for JSON representation size of our test networks).

### CODE REVIEW

File: alchemiscale/tests/integration/interface/test_api.py [hunk 1]
```diff
@@ -6,6 +6,9 @@
 
 from alchemiscale.models import ScopedKey
 from alchemiscale.base.client import json_to_gufe
+from alchemiscale.utils import keyed_dicts_to_gufe, gufe_to_keyed_dicts
+
+import networkx as nx
```
Comment (dotsdl) [inline alchemiscale/tests/integration/interface/test_api.py:11]: I don't think this import is used in this module.

**Review by dotsdl — CHANGES_REQUESTED** [review]
Looking fantastic @ianmkenney! This is a huge improvement!

Can you also add a unit test (not integration test) for each of the methods this PR adds in `alchemiscale.utils`? We want to be sure in a fairly isolated way that any `GufeTokenizable`s these methods handle are working as expected via these unit tests.

File: alchemiscale/interface/client.py [hunk 2]
```diff
@@ -17,6 +17,7 @@
 from gufe import AlchemicalNetwork, Transformation, ChemicalSystem
 from gufe.tokenization import GufeTokenizable, JSON_HANDLER, GufeKey
 from gufe.protocols import ProtocolResult, ProtocolDAGResult
+import networkx as nx
```
[the remark this answers is not in the document]
Comment (ianmkenney) [reply alchemiscale/interface/client.py:20]: seems we do, see `get_transformation_tasks`

[the remark this answers is not in the document]
Comment (dotsdl) [reply alchemiscale/interface/client.py:20]: Oh that's strange; not sure how we didn't already have this import :thinking:

[the remark this answers is not in the document]
Comment (ianmkenney) [reply alchemiscale/interface/client.py]: we did, I added it as a repeated import because I didn't notice it 😮‍💨

**Review by dotsdl — APPROVED** [review]
Amazing work @ianmkenney! Merging!

**Pull request merged**